native/src/init/getinfo.cpp

```diff
@@ -171,6 +171,45 @@ if (access(file_name, R_OK) == 0) {                                 \
     }                                                               \
 }
 
+
+// About `partition_map`, `load_partition_map` and `get_partition_name_for_device`:
+// `androidboot.partition_map` allows associating a partition name for a raw block device
+// through a comma separated and semicolon deliminated list. For example,
+// `androidboot.partition_map=vdb,metadata;vdc,userdata` maps `vdb` to `metadata` and `vdc` to
+// `userdata`.
+// https://android.googlesource.com/platform/system/core/+/refs/heads/android13-release/init/devices.cpp#191
+
+kv_pairs partition_map;
+
+void load_partition_map() {
+    auto fn = [](const kv_pairs &kv) {
+        for (const auto &[key, value] : kv) {
+            if (key == "androidboot.partition_map") {
+                for (const auto& map : split(value, ";")) {
+                    auto map_pieces = split(map, ",");
+                    if (map_pieces.size() != 2) {
+                        LOGE("Expected a comma separated device,partition mapping, but found '%s'", map.c_str());
+                        continue;
+                    }
+                    partition_map.emplace_back(map_pieces[0], map_pieces[1]);
+                }
+            }
+        }
+    };
+ 
+    fn(parse_cmdline(full_read("/proc/cmdline")));
+    fn(parse_bootconfig(full_read("/proc/bootconfig")));   
```

Comment:
why not add `androidboot.partition_map` to `BootConfig::set` and use `Config` here?

Reply:
`BootConfig::set` is only being called from `load_kernel_info`, which doesn't be called in the SecondStageInit

Reply:
Then why not make partition_map a `SecondStageInit`'s member

Reply:
`LegacySARInit`, `RootFSInit` and `SecondStageInit` call the `setup_block()`. 

It might be worth extending the scope of `load_partition_map`, or adding some comments to make it clear that it was added as a hack for the Google Play Games Developer Emulator.

 I think we should go for the extended scope approach, which will reduce the code paths.

Reply:
Then this is not good. Just make `partition_map` a local variable of `setup_block`, since you only use `get_partition_name_for_device` there. I think it's safe for other `Init`s to support this fallback.

Reply:
the `partition_map` will never change, I don't see the point of make it a local variable.

Reply:
It's not about mutable or not.